node/src/components/storage/object_pool.rs

```diff
@@ -0,0 +1,180 @@
+//! A reference pool for items/objects.
+//!
+//! Its core responsibility is to deduplicate potentially expensive loads by keeping a weak
+//! reference to any loaded object around, so that any load request for an object that is currently
+//! in active use can be satisfied using the already existing copy.
+//!
+//! It differs from a cache in that it does not hold strong references to an item itself -- once an
+//! item is no longer used, it will not be kept in the pool for a later request. As a consequence
+//! the memory pool will never consume significantly more memory than what would otherwise be
+//! required by the loaded objects that are in active use anyway and thus has an "infinite"
+//! capacity.
+use std::{
+    borrow::Borrow,
+    collections::HashMap,
+    hash::Hash,
+    sync::{Arc, Weak},
+};
+
+use datasize::DataSize;
+
+/// A pool of items/objects.
+///
+/// Maintains a pool of weak references and automatically purges them in configurable intervals.
+#[derive(DataSize, Debug)]
+pub(super) struct ObjectPool<I> {
+    /// The actual object pool.
+    #[data_size(skip)]
```

Comment:
But maybe the object pool would be the right place to count these? If we ignore the `Arc`s elsewhere, they won't be double-counted, and there's always only one reference in the object pool.

Reply:
That's a pretty decent point. I'll see if I can add a manual `impl` for this reason.